### ISSUE TITLE
Button actions from CollectionViewCells

### ISSUE THREAD

**Issue by olafurr** [ReactiveCocoa/ReactiveCocoa#1675]
I'm working on an app which has an "infinite scrolling" UICollectionView with a couple of cells per section. Some of the cells have a button which will in turn cause insertions or deletions from the CV and will fire of network requests. What I'm trying to do is to make the viewModel subscribe to signals from the button action in the cell which will fire off a network request and then forward the signal to the controller so it can update it's CV. What I'm doing right now is this:

**CustomCollectionViewCell.h**
```objective-c
@property (strong, nonatomic) RACSignal *clickedButtonSignal;
```

**CustomCollectionViewCell.m**
```objective-c
- (IBAction)buttonClicked {
         // Already initialized the clickedButtonSignal elsewhere
         [(RACSubject *)self.clickedButtonSignal sendNext:self.indexPath];
}
```

**CollectionViewDataSource.m**
```objective-c
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
        // Dequeue cell
        cell.clickedButtonSignal = self.viewModel.clickedButtonSignal;
}
```

**ViewModel.h**
```objective-c
@property (strong, nonatomic) RACSignal *clickedButtonSignal;
@property (strong, nonatomic) RACSignal *insertSectionsSignal;
```

**ViewModel.m**
```objective-c
- (id)init {
      // Doing stuff
    
      [self.clickedButtonSignal subscribeNext:^(NSIndexPath *indexPath) {
             // Get the correct item and make a network request
            
             // The controller subscribes to this signal and updates the CV respectively
             [(RACSubject *)self.insertSectionsSignal sendNext:[NSValue valueWithRange:NSMakeRange()];
      }];
}
```

So basically each CVCell has a reference to the clickedButtonSignal in the viewModel (which just doesn't seem right).

My other idea was that the clickedButtonSignal in the cells would would use rac_signalForSelector: and then the viewModel would observe all those signal but with maybe 150 cells in the CV I started thinking that could be bad performance wise.

Right now my solution is working but like I said it just doesn't seem right.

I'd be very greatful if anyone could comment on my solution so far or give me any ideas on how I could implement this.

Thanks in advance!

**Comment by Tiger6688** [comment]
less functional and more imperative,It seems as if there is no need to use reactiveCocoa,there is no composing and transforming streams of values.

**Comment by ReDetection** [comment]
@olafurr consider using `RACCommand` instead of `clickedButtonSignal` and make `insertSectionsSignal` to depend on `clickedButtonSignal` in declarative way like so:
```
- (id)init {
      // Doing stuff

      self.insertSectionsSignal = [[self.clickedButtonSignal 
          map:^id(id x) { //or map or anything
                // get the correct item
          }]
          performNetworkRequest];
}
```
Also learn `@weakify()` and `@strongify()` macro to avoid retain cycles and memory leaks.
Happy reactive programming!

**Comment by olafurr** [comment]
Thanks, I will try that. I just thought it would be too much overhead to have a RACCommand for each cell if say I have 200 cells in the CollectionView.

I also stumbled upon rac_signalForSelector:fromProtocol: which works quite nicely.

**Comment by ReDetection** [comment]
Keep only one in the viewModel and assign it to the every cell's button. Also note that you don't need to reassign command if button was configured by you before and just dequeued from unused cells queue. So I see no overhead in this case.

**Comment by ahoang** [comment]
If I understand your question correctly, you want to observe when a button in one of your collection view cells are pressed and respond to that button press by potentially firing off a network request and then updating (inserting, deleting) your collection view?

If your collectionView cell has a UIButton, set the button's rac_command to a rac_command in the viewModel. When the button is pressed, it will call the RACCommand's 'execute:', and do whatever work you desire in the view model. When that work is finished in the view model, update the data in the view model. Your view controller should be observing changes to the view model's data.

Your view model:
<pre><code>
@interface ViewModel : NSObject

// This is the data that backs your collection view
@property (nonatomic, strong) NSArray *data;
@property (nonatomic, strong, readonly) RACCommand *cellButtonCommand;

@end
</code></pre>

<pre><code>

@implementation ViewModel

- (RACCommnad *)cellButtonCommand {

   if(self.cellButtonCommand) {
     return self.cellButtonCommand;
  }
  
@weakify(self)
 self.cellButtonCommand = [[RACCommand alloc] initWithSignal:^RACSIgnal *(id x) {
@strongify(self)
// Do your work/network request here
// Update your data
self.data = newData;

}];

}


@end

</code></pre>

In your ViewController's viewDidLoad

<pre><code>

- (void)viewDidLoad {
   [super viewDidLoad];

   // The data that backs your collection view should bind to the view model's data.
   RAC(self, data) = RACObserve(self, viewModel.data);

   // Observe changes to the view model's data and reload the collection view
   @weakify(self)
   [RACObserve(self, viewModel.data) subscribeNext:^(NSArray *data) {

       @strongify(self)
       [self.collectionView reloadData];

     }];
}

</code></pre>

And in your cellForRow

<pre><code>

cell.button.rac_command = self.viewModel.cellButtonCommand;

</code></pre>

**Comment by ReDetection** [comment]
@ahoang wtf man, you're calling getter inside getter, this will lead to a loop until stack overflow. Also `RACCommand`, not `RACCommnad`
```
- (RACCommnad *)cellButtonCommand {

   if(self.cellButtonCommand) {
     return self.cellButtonCommand;
  }
```

**Comment by ahoang** [comment]
Whoops. wrote it in plain text so i'm sure i fucked up

**Comment by RuiAAPeres** [comment]
Closing due to inactivity.
